### ISSUE TITLE
Cannot get beyond block 99550000

### ISSUE THREAD

**Issue by jseibel-circle** [binance-chain/node-binary#222]
We have upgraded to latest binary 0.7.2. We updated our `app.toml` to the latest. We have tried `state_recover` at multiple different heights. Still, always hang on 99550000. Same issue with multiple nodes that have independent blockchain storage.

Error:
```
ABCI Handshake App Info module=consensus height=99550000 hash=16420A2203305D86BB6AAE985CD8E66FFF28E16D047FD5E0FE18A9054172A1B4 software-version= protocol-version=0
ABCI Replay Blocks module=consensus appHeight=99550000 storeHeight=99550001 stateHeight=99550000
Replay last block using real app module=consensus
ERROR: error during handshake: error on replay: Wrong Block.Header.AppHash. Expected 16420A2203305D86BB6AAE985CD8E66FFF28E16D047FD5E0FE18A9054172A1B4, got F269C5089D10AC2D6AAC6CE27AF34B95166B4EEC24AE1705D838F9265D0531A5
INFO exited: bnb-daemon (exit status 1; not expected)
```
We have tried recovering 1 block before:
`./state_recover 99549999 <blockchain folder>`
10k before:
`./state_recover 99540000  <blockchain folder>`
30k before:
`./state_recover 99520000  <blockchain folder>`

They get to 99550000 and then errors. What are we doing wrong? How can we fix this? Thank you

**Comment by huangsuyu** [comment]
from your log
```
storeHeight=99550001
```
it seems the recover is not successful, please check your recover command again

**Comment by jseibel-circle** [comment]
Our full logs from recovery to error.

Notice how `ABCI Replay Blocks module=consensus appHeight=99549999 storeHeight=99549999 stateHeight=99549999` is logged correctly at the beginning, but when it crashes at the end it has since gotten to that problem again.

```
$ ./state_recover 99549999 <our blockchain folder>
recover home_path[/data] to height[99549999]
recover success

< we start bnb-daemon >
bnb-daemon: started

...
< The logs >

ABCI Replay Blocks module=consensus appHeight=99549999 storeHeight=99549999 stateHeight=99549999

... a lot of logs here ...

Added Order module=dex order="NewOrderMsg{Sender: 6E54A939CAF77F2E2817AD733D317CFC5FC33698, Id: 6E54A939CAF77F2E2817AD733D317CFC5FC33698-1110630, Symbol: BTCB-1DE_USDSB-1AC}"
Canceled Order module=dex order="CancelOrderMsg{Sender:10F1A69943FE613E7291F3D230A4F8E3EEA4AEA0, RefId: 10F1A69943FE613E7291F3D230A4F8E3EEA4AEA0-1370808}"
replayed all tx. Starting match module=dex height=99549996
Relaying block for order book module=dex height=99549997
Added Order module=dex order="NewOrderMsg{Sender: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D, Id: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D-47786, Symbol: AVA-645_BUSD-BD1}"
Canceled Order module=dex order="CancelOrderMsg{Sender:6E54A939CAF77F2E2817AD733D317CFC5FC33698, RefId: 6E54A939CAF77F2E2817AD733D317CFC5FC33698-1110629}"
replayed all tx. Starting match module=dex height=99549997
Relaying block for order book module=dex height=99549998
Added Order module=dex order="NewOrderMsg{Sender: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D, Id: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D-47787, Symbol: BNB_BUSD-BD1}"
Canceled Order module=dex order="CancelOrderMsg{Sender:1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E, RefId: 1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E-8773090}"
Canceled Order module=dex order="CancelOrderMsg{Sender:6E54A939CAF77F2E2817AD733D317CFC5FC33698, RefId: 6E54A939CAF77F2E2817AD733D317CFC5FC33698-1110630}"
Canceled Order module=dex order="CancelOrderMsg{Sender:D3D7974E0E537FFECF6CADD8C6E38BFB98D3B37A, RefId: D3D7974E0E537FFECF6CADD8C6E38BFB98D3B37A-370623}"
replayed all tx. Starting match module=dex height=99549998
Relaying block for order book module=dex height=99549999
Added Order module=dex order="NewOrderMsg{Sender: EEE14CF1EB65860C823B00AD4665EBBDBD05FCA8, Id: EEE14CF1EB65860C823B00AD4665EBBDBD05FCA8-3, Symbol: BNB_ETH-1C9}"
Canceled Order module=dex order="CancelOrderMsg{Sender:88E78538A4AC1EA4BAECBD46769B576B80C6DC29, RefId: 88E78538A4AC1EA4BAECBD46769B576B80C6DC29-1611786}"
replayed all tx. Starting match module=dex height=99549999
Relaying block for order book module=dex height=99550000
Canceled Order module=dex order="CancelOrderMsg{Sender:1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E, RefId: 1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E-8773118}"
replayed all tx. Starting match module=dex height=99550000
No order comes in for the block module=dexkeeper
Located day to load breathe block height module=dexkeeper epochDay="\u0014H\u0000\u0000\u0000\u0000\u0000\u0000"
Loaded breathe block height module=dexkeeper height=99530843
get last breathe block height module=main height=99530843
reload latest snapshot module=statesync height=99530843
ABCI Handshake App Info module=consensus height=99550000 hash=16420A2203305D86BB6AAE985CD8E66FFF28E16D047FD5E0FE18A9054172A1B4 software-version= protocol-version=0
ABCI Replay Blocks module=consensus appHeight=99550000 storeHeight=99550001 stateHeight=99550000
Replay last block using real app module=consensus
ERROR: error during handshake: error on replay: Wrong Block.Header.AppHash. Expected 16420A2203305D86BB6AAE985CD8E66FFF28E16D047FD5E0FE18A9054172A1B4, got F269C5089D10AC2D6AAC6CE27AF34B95166B4EEC24AE1705D838F9265D0531A5
INFO exited: bnb-daemon (exit status 1; not expected)
```

**Comment by forcodedancing** [comment]
> Our full logs from recovery to error.
> 
> Notice how `ABCI Replay Blocks module=consensus appHeight=99549999 storeHeight=99549999 stateHeight=99549999` is logged correctly at the beginning, but when it crashes at the end it has since gotten to that problem again.
> 
> ```
> $ ./state_recover 99549999 <our blockchain folder>
> recover home_path[/data] to height[99549999]
> recover success
> 
> < we start bnb-daemon >
> bnb-daemon: started
> 
> ...
> < The logs >
> 
> ABCI Replay Blocks module=consensus appHeight=99549999 storeHeight=99549999 stateHeight=99549999
> 
> ... a lot of logs here ...
> 
> Added Order module=dex order="NewOrderMsg{Sender: 6E54A939CAF77F2E2817AD733D317CFC5FC33698, Id: 6E54A939CAF77F2E2817AD733D317CFC5FC33698-1110630, Symbol: BTCB-1DE_USDSB-1AC}"
> Canceled Order module=dex order="CancelOrderMsg{Sender:10F1A69943FE613E7291F3D230A4F8E3EEA4AEA0, RefId: 10F1A69943FE613E7291F3D230A4F8E3EEA4AEA0-1370808}"
> replayed all tx. Starting match module=dex height=99549996
> Relaying block for order book module=dex height=99549997
> Added Order module=dex order="NewOrderMsg{Sender: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D, Id: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D-47786, Symbol: AVA-645_BUSD-BD1}"
> Canceled Order module=dex order="CancelOrderMsg{Sender:6E54A939CAF77F2E2817AD733D317CFC5FC33698, RefId: 6E54A939CAF77F2E2817AD733D317CFC5FC33698-1110629}"
> replayed all tx. Starting match module=dex height=99549997
> Relaying block for order book module=dex height=99549998
> Added Order module=dex order="NewOrderMsg{Sender: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D, Id: 0E7139E7EC6A21B0ED17D2CD33BE5D8C499A857D-47787, Symbol: BNB_BUSD-BD1}"
> Canceled Order module=dex order="CancelOrderMsg{Sender:1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E, RefId: 1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E-8773090}"
> Canceled Order module=dex order="CancelOrderMsg{Sender:6E54A939CAF77F2E2817AD733D317CFC5FC33698, RefId: 6E54A939CAF77F2E2817AD733D317CFC5FC33698-1110630}"
> Canceled Order module=dex order="CancelOrderMsg{Sender:D3D7974E0E537FFECF6CADD8C6E38BFB98D3B37A, RefId: D3D7974E0E537FFECF6CADD8C6E38BFB98D3B37A-370623}"
> replayed all tx. Starting match module=dex height=99549998
> Relaying block for order book module=dex height=99549999
> Added Order module=dex order="NewOrderMsg{Sender: EEE14CF1EB65860C823B00AD4665EBBDBD05FCA8, Id: EEE14CF1EB65860C823B00AD4665EBBDBD05FCA8-3, Symbol: BNB_ETH-1C9}"
> Canceled Order module=dex order="CancelOrderMsg{Sender:88E78538A4AC1EA4BAECBD46769B576B80C6DC29, RefId: 88E78538A4AC1EA4BAECBD46769B576B80C6DC29-1611786}"
> replayed all tx. Starting match module=dex height=99549999
> Relaying block for order book module=dex height=99550000
> Canceled Order module=dex order="CancelOrderMsg{Sender:1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E, RefId: 1468EE412C3ADC9CFF3EF31ADC7EDD288F5E208E-8773118}"
> replayed all tx. Starting match module=dex height=99550000
> No order comes in for the block module=dexkeeper
> Located day to load breathe block height module=dexkeeper epochDay="\u0014H\u0000\u0000\u0000\u0000\u0000\u0000"
> Loaded breathe block height module=dexkeeper height=99530843
> get last breathe block height module=main height=99530843
> reload latest snapshot module=statesync height=99530843
> ABCI Handshake App Info module=consensus height=99550000 hash=16420A2203305D86BB6AAE985CD8E66FFF28E16D047FD5E0FE18A9054172A1B4 software-version= protocol-version=0
> ABCI Replay Blocks module=consensus appHeight=99550000 storeHeight=99550001 stateHeight=99550000
> Replay last block using real app module=consensus
> ERROR: error during handshake: error on replay: Wrong Block.Header.AppHash. Expected 16420A2203305D86BB6AAE985CD8E66FFF28E16D047FD5E0FE18A9054172A1B4, got F269C5089D10AC2D6AAC6CE27AF34B95166B4EEC24AE1705D838F9265D0531A5
> INFO exited: bnb-daemon (exit status 1; not expected)
> ```

From the logs, we can see the upgrade did not take effects on your side.
Please double check you are using the right app.toml (review upgrade height carefully), and it is in the correct location.

https://github.com/binance-chain/node-binary/blob/master/fullnode/prod/0.7.2/config/app.toml

If you still have issues, please share your app.toml.

**Comment by jseibel-circle** [comment]
We diff'd that app.toml with outs. Only difference is `logToConsole = false` changed to `logToConsole = true`

We indeed took your app.toml when you published it, as we always do.

**Comment by jseibel-circle** [comment]
We got past the block by nuking our blockchain and resyncing from the tip. We have a large gap of blocks now and no longer have our history, but at least we got past the block.

We have had this issue numerous times (`Wrong Block.Header.AppHash`), as do many other people who have posted similar issues. I recommend enhancing your debugging output, as this is a closed-source blockchain so we cannot read the code, unlike most other blockchains. The errors are mystifying and make it impossible for us to know what to do, or for you to help us.
